homeassistant/components/dreo/fan.py

```diff
@@ -0,0 +1,195 @@
+
+    def __init__(
+        self,
+        device: dict[str, Any],
+        coordinator: DreoDataUpdateCoordinator,
+    ) -> None:
+        """Initialize the Dreo fan."""
+        super().__init__(device, coordinator, UNIQUE_ID_SUFFIX, None)
+
+        model_config = FAN_DEVICE.get("config", {}).get(self._model, {})
+        speed_range = model_config.get("speed_range")
+
+        self._attr_preset_modes = model_config.get("preset_modes")
+        self._low_high_range = speed_range
+
+        self._update_attributes()
+
+    @callback
+    def _handle_coordinator_update(self):
+        """Handle updated data from the coordinator."""
+        self._update_attributes()
+        super()._handle_coordinator_update()
+
+    def _update_attributes(self):
+        """Update attributes from coordinator data."""
+        if not self.coordinator.data:
+            self._attr_available = False
+            return
+
+        fan_state_data = self.coordinator.data
+        if fan_state_data.available is None:
+            self._attr_available = False
+            return
+
+        self._attr_available = fan_state_data.available
+
+        if not fan_state_data.is_on:
+            self._attr_percentage = 0
+            self._attr_preset_mode = None
+            self._attr_oscillating = None
+        else:
+            self._attr_preset_mode = fan_state_data.mode
+            self._attr_oscillating = fan_state_data.oscillate
+            self._attr_percentage = fan_state_data.speed_percentage
+
+    async def async_turn_on(
+        self,
+        percentage: int | None = None,
+        preset_mode: str | None = None,
+        **kwargs: Any,
+    ) -> None:
+        """Turn the device on."""
+        await self.async_execute_fan_common_command(
+            ERROR_TURN_ON_FAILED, percentage=percentage, preset_mode=preset_mode
+        )
+
+    async def async_turn_off(self, **kwargs: Any) -> None:
+        """Turn the device off."""
+        await self.async_send_command_and_update(
+            ERROR_TURN_OFF_FAILED, power_switch=False
+        )
+
+    async def async_set_preset_mode(self, preset_mode: str) -> None:
+        """Set the preset mode of fan."""
+        await self.async_execute_fan_common_command(
+            ERROR_SET_PRESET_MODE_FAILED, preset_mode=preset_mode
+        )
+
+    async def async_set_percentage(self, percentage: int) -> None:
+        """Set the speed of fan."""
+
+        if percentage <= 0:
+            await self.async_turn_off()
+            return
+
+        await self.async_execute_fan_common_command(
+            ERROR_SET_SPEED_FAILED, percentage=percentage
+        )
+
+    async def async_oscillate(self, oscillating: bool) -> None:
+        """Set the oscillation of fan."""
+        await self.async_execute_fan_common_command(
+            ERROR_SET_OSCILLATE_FAILED, oscillate=oscillating
+        )
+
+    async def async_execute_fan_common_command(
+        self,
+        translation_key: str,
```

Comment:
```suggestion
        error_translation_key: str,
```
this should be consistent with `async_send_command_and_update`